### DISCUSSION TITLE
Geometry Shaders in Stride3D

### DISCUSSION THREAD

**Discussion by najak3d** [stride3d/stride#1307] [Q&A]
I saw that Geometry Shaders seem to be supported by Stride3D.   I'm posting here just to confirm this reality.

In our app, we use LOTS of Vector data loaded and unloaded often.    One thing we have is a nice implementation of outlined line data, and achieve this via created 2-3 Triangles per line segment (3rd triangle creates the join, for turns).   Which means, each time new line data loads up, on the CPU we've got to build this mesh.

What we'd prefer is to just have it done ONCE by the Geometry Shader on first load of each Line, which would do the same work we're already doing in C#, but would use the GPU to create the VertexData for the Vertex Shader.

Since Urho.NET does not support Geometry Shaders, it's been a long time since I've used these.   Is it feasible to just have the Geometry Shader run ONCE for a given PolyLine, and from then on, only run the Vertex/Pixel Shaders?

Our understanding is that GeometryShaders often incur a large performance hit.  And from our perspective, the Geometry shader really only needs to be run ONCE (1st frame) for each new PolyLine/Shape that we load.

If instead we are forced to re-run the Geometry Shader EVERY FRAME (i.e. it won't cache the resulting VertexBuffer for us) -- then we won't bother with Geometry Shaders.   But if we can make it only render ONCE (per PolyLine), then this becomes a very appealing option to us  (i.e. Saves CPU load, and RAM usage).

Here's a snapshot of our App, showing a sample of the lines renderered - for Roads, Rivers, State Lines, Routes, Airspaces, etc.

![image](https://user-images.githubusercontent.com/10914515/151684614-4af87746-0b14-4e31-9f35-b9500e8c8a88.png)

**Comment by Eideren** [comment]
Isn't the geometry shader part of the rendering stage, and so must run every frame to render, it wouldn't cache anything, like vertex and fragment shaders.
Sounds like you'll have to prepare those buffers on cpu or use a compute shader instead.

**Reply by najak3d** [reply]
How do you use "Compute Shader" with Stride3D?   What I need to achieve is sending "just the raw points to the GPU Compute Shader" and then having it create the VertexBuffer to be used by my Vertex Shader directly, and avoid sending all that VertexData back to the CPU, only to be re-blitted back to the GPU.

If this can be done, it also enables us to do more heavy lifting that we do now for Calculating Real-Earth Terrain Elevation Normals -- right now we do this massively on the CPU (as our database doesn't have Normals stored in it, as it would more than double the size of our Elevations DB (more RAM, more disk space, more to read from disk).     For Elevation Normals, we would want to use our Compute Shader to render it's results directly to a Texture, that will be used by Pixel Shader... but we DO NOT WANT to regenerate these Normals EVERY FRAME (they NEVER change).

So if we can get the Compute Shader to process our data, and allow us to reuse this data by another shader, without it ever coming back to the CPU -- OMG, awesome!

I'd like to find out how to accomplish this via Stride3D.   Urho.NET offers no support for Geometry Shaders, so this would be a competitive advantage of Stride3D over Urho3D.

**Reply by Eideren** [reply]
There's a couple of examples in the source, [this one seems like the most straightforward](https://github.com/stride3d/stride/blob/master/sources/engine/Stride.Graphics.Tests.11_0/TestComputeShader.cs), but to be honest it's probably less of a hassle to generate them on CPU and send them to the GPU as a plain old vertex buffer, stride won't keep that buffer in cpu memory if you don't.
If the data is too large to store in gpu memory then just rely on the geometry shader and skip it when the geometry is not on screen. If your output mesh is expected to be less than 10000 triangles you really shouldn't worry about running it every frame.

**Reply by najak3d** [reply]
I'm going to ask this as a new question now, so that it can be searchable by those looking for this answer.

If you'll re-answer it here, then you can get the proper credit for supplying this answer.

https://github.com/stride3d/stride/discussions/1309
